com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ModuleHooksAggregator.java

```diff
@@ -0,0 +1,145 @@
+        this.moduleToDeploy = moduleToDeploy;
+    }
+
+    public List<Hook> aggregateHooks(HookPhase currentHookPhaseForExecution) {
+        substituteBeforeStartHooksPhases();
+        Map<String, List<String>> alreadyExecutedHooksForModule = getAlreadyExecutedHooks();
+        List<Hook> hooksCalculatedForExecution = determineHooksForExecution(alreadyExecutedHooksForModule, currentHookPhaseForExecution);
+        updateExecutedHooksForModule(alreadyExecutedHooksForModule, currentHookPhaseForExecution, hooksCalculatedForExecution);
+        return hooksCalculatedForExecution;
+    }
+
+    private void substituteBeforeStartHooksPhases() {
+        for (Hook hook : getModuleHooks(moduleToDeploy)) {
+            List<String> phases = new ArrayList<>(hook.getPhases());
+            if (phases.contains(HookPhase.APPLICATION_BEFORE_START.getValue())) {
+                phases.remove(HookPhase.APPLICATION_BEFORE_START.getValue());
+                phases.add(HookPhase.APPLICATION_BEFORE_START_LIVE.getValue());
+                phases.add(HookPhase.APPLICATION_BEFORE_START_IDLE.getValue());
+                hook.setPhases(phases);
+            }
+        }
+    }
+
+    private Map<String, List<String>> getAlreadyExecutedHooks() {
+        return StepsUtil.getExecutedHooksForModule(execution, moduleToDeploy.getName());
+    }
+
+    private List<Hook> determineHooksForExecution(Map<String, List<String>> alreadyExecutedHooks, HookPhase hookPhaseForCurrentStepPhase) {
+        List<Hook> moduleHooksToExecuteOnCurrentStepPhase = collectHooksWithPhase(moduleToDeploy, hookPhaseForCurrentStepPhase);
+        return getHooksForExecution(alreadyExecutedHooks, moduleHooksToExecuteOnCurrentStepPhase, hookPhaseForCurrentStepPhase);
+    }
+
+    private List<Hook> collectHooksWithPhase(Module moduleToDeploy, HookPhase hookTypeForCurrentStepPhase) {
+        return getModuleHooks(moduleToDeploy).stream()
+                                             .filter(hook -> shouldExecuteHook(hook.getPhases(), hookTypeForCurrentStepPhase))
+                                             .collect(Collectors.toList());
+    }
+
+    private List<Hook> getModuleHooks(Module moduleToDeploy) {
+        return moduleToDeploy.getMajorSchemaVersion() < MAJOR_SCHEMA_VERSION_THREE ? Collections.emptyList() : moduleToDeploy.getHooks();
+    }
+
+    private boolean shouldExecuteHook(List<String> hookPhases, HookPhase hookTypeForCurrentStepPhase) {
+        List<HookPhase> resolvedHookPhases = mapToHookPhases(hookPhases);
+        return resolvedHookPhases.contains(hookTypeForCurrentStepPhase);
+    }
+
+    private List<HookPhase> mapToHookPhases(List<String> hookPhases) {
+        return hookPhases.stream()
+                         .map(HookPhase::fromString)
+                         .collect(Collectors.toList());
+    }
+
+    private List<Hook> getHooksForExecution(Map<String, List<String>> alreadyExecutedHooks, List<Hook> moduleHooksToBeExecuted,
```

Comment:
This method can be done with one `stream().filter().collect()` chain.
<pre> hook -> !alreadyExecutedHooks.containsKey(hook.getName()) || !hasNonExecutedPhases(alreadyExecutedHooks, hook, hookPhaseForCurrentStepPhase)</pre>
That way, there is no need for the following 2 methods